x-pack/platform/plugins/shared/alerting/server/maintenance_window_events/task.ts

```diff
@@ -0,0 +1,296 @@
+      sortField: 'updatedAt',
+      sortOrder: 'desc',
+      filter,
+    });
+  } catch (e) {
+    logger.error(`Failed to find maintenance windows saved object". Error: ${e.message}`);
+    return null;
+  }
+}
+
+export async function generateEvents({
+  maintenanceWindowsSO,
+  startRangeDate,
+}: {
+  maintenanceWindowsSO: Array<SavedObject<MaintenanceWindowAttributes>>;
+  startRangeDate: string;
+}) {
+  const maintenanceWindows = maintenanceWindowsSO.map((savedObject) =>
+    transformMaintenanceWindowAttributesToMaintenanceWindow({
+      attributes: savedObject.attributes,
+      id: savedObject.id,
+    })
+  );
+
+  const newExpirationDate = moment(startRangeDate).utc().add(1, 'year').toISOString();
+
+  const mwWithGeneratedEvents = maintenanceWindows
+    // filtering the maintenance windows that have recurring schedule and events
+    .filter(
+      (maintenanceWindow) =>
+        (maintenanceWindow.rRule.interval !== undefined ||
+          maintenanceWindow.rRule.freq !== undefined) &&
+        maintenanceWindow.events.length
+    )
+    .map((filteredMaintenanceWindow) => {
+      const { rRule, duration, expirationDate: oldExpirationDate } = filteredMaintenanceWindow;
+
+      const newEvents = generateMaintenanceWindowEvents({
+        rRule,
+        expirationDate: newExpirationDate,
+        duration,
+        startDate: startRangeDate, // here start range date is 1 week before current expiration date
+      });
+
+      return {
+        ...filteredMaintenanceWindow,
+        expirationDate: newEvents.length ? newExpirationDate : oldExpirationDate,
+        events: [...newEvents],
+      };
+    })
+    .filter((mappedMW) => mappedMW.expirationDate === newExpirationDate);
+
+  return mwWithGeneratedEvents;
+}
+
+export const updateMaintenanceWindowsEvents = async ({
+  filter,
+  savedObjectsClient,
+  logger,
+  startRangeDate,
+}: {
+  logger: Logger;
+  savedObjectsClient: ISavedObjectsRepository;
+  filter: KueryNode;
+  startRangeDate: string;
+}) => {
+  let totalUpdatedMaintenanceWindows = 0;
+  let mwsWithNewEvents = [];
+  const soFinder = await getSOFinder({
+    savedObjectsClient,
+    logger,
+    filter,
+  });
+
+  if (soFinder) {
+    for await (const findResults of soFinder.find()) {
+      mwsWithNewEvents = await generateEvents({
+        maintenanceWindowsSO: findResults.saved_objects,
+        startRangeDate,
+      });
+
+      try {
+        if (mwsWithNewEvents.length) {
+          const bulkUpdateReq = mwsWithNewEvents.map((mw) => {
+            const updatedMaintenanceWindowAttributes =
+              transformMaintenanceWindowToMaintenanceWindowAttributes({
+                ...mw,
+              });
+
+            return {
+              type: MAINTENANCE_WINDOW_SAVED_OBJECT_TYPE,
+              id: mw.id,
+              attributes: updatedMaintenanceWindowAttributes,
+            };
+          });
+
+          const result = await savedObjectsClient.bulkUpdate<MaintenanceWindowAttributes>(
```

Comment:
Some of the objects in the `results.saved_objects` may contain an `error` object. Should we handle it? I think we have two options: a) ignore errors, and on the next run of the task, they will probably get updated correctly, and b) throw an error and let the task manager retry the task. Wdyt?

Reply:
Task manager retries recurring task on the next schedule run anyway, so if saved object had an error and is not updated, I think it should probably be retried again next day.

Reply:
I think logging any individual errors in the bulk list makes sense.  Seems like there are a bunch of options for retry-ing this we may want to explore later:

- framework handling retries for recurring tasks (currently we don't, per Ying's comment ^^^)
- retry within the task a few times
- somehow launch an adhoc/non-recurring version of this task, specifically for retry purposes (till the framework can do this)

In any case, by the time the task ends, I'd expect any errors it couldn't fix to be logged.